flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullMaskUtils.scala

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.flink.api.table.typeinfo
+
+import org.apache.flink.api.table.Row
+import org.apache.flink.core.memory.{DataInputView, DataOutputView}
+
+object NullMaskUtils {
+
+  def writeNullMask(len: Int, value: Row, target: DataOutputView): Unit = {
+    var b = 0x00
+    var bytePos = 0
+
+    var fieldPos = 0
+    while (fieldPos < len) {
+      b = 0x00
+      // set bits in byte
+      bytePos = 0
+      while (bytePos < 8 && fieldPos < len) {
+        b = b << 1
+        // set bit if field is null
+        if(value.productElement(fieldPos) == null) {
+          b |= 0x01
+        }
+        bytePos += 1
+        fieldPos += 1
+      }
+      // shift bits if last byte is not completely filled
+      while (bytePos < 8) {
+        b = b << 1
+        bytePos += 1
+      }
```

Comment:
If you're really concerned about performance, then this should be `b <<= (8 - bytePos)`.